File: android/app/src/test/java/ai/oriveo/community/core/provider/ProductionCapabilitySnapshotTest.kt
package ai.oriveo.community.core.provider

import ai.oriveo.community.core.data.remote.MetadataClient
import ai.oriveo.community.core.model.AIModel
import ai.oriveo.community.core.model.Provider
import ai.oriveo.community.core.model.ProviderKind
import java.io.File
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Drives the real decision function [CapabilityControlResolution.resolve] with a slice of
 * **actually published** metadata rather than yet another set of hand-written fixtures.
 *
 * Why not hand-written fixtures: an earlier round was verified entirely by "unit tests pass,
 * the curl output has the right fields, the deploy returned 200". None of those three say
 * anything about what a user sees on screen, and every one of those tests had built its own
 * data specifically so that its assertion would hold. The state and profile combinations here
 * are lifted verbatim out of the published catalog, and the expected values live in the same
 * slice (`expectedVerdicts`) so that they cannot be quietly rewritten to match the code.
 */
class ProductionCapabilitySnapshotTest {
    private val json = Json { ignoreUnknownKeys = true }

    private val snapshot: JsonObject by lazy {
        json.parseToJsonElement(
            workspaceFile("shared/model-contracts/production-capability-snapshot.json").readText(),
        ).jsonObject
    }

    private val models: Map<String, JsonObject> by lazy {
        snapshot["models"]!!.jsonObject.mapValues { it.value.jsonObject }
    }

    private val expectedVerdicts: Map<String, JsonObject> by lazy {
        snapshot["expectedVerdicts"]!!.jsonObject
            .filterKeys { !it.startsWith("$") }
            .mapValues { it.value.jsonObject }
    }

    /**
     * The registry itself is used as the capabilityRuntime, not a mock of it.
     *
     * A false green of exactly this shape was fixed once before: the mock supplied
     * `recipes: {}`, so every auto_available control carrying a recipeRef fell through
     * `dangling_recipe_ref` into the legacy path, and the assertion claiming "the legacy
     * profile is not consulted" was in fact passing because of the legacy profile. So this
     * reads the registry file directly instead of inventing a recipe set.
     */
    private fun metadataClient(): MetadataClient {
        val registry = workspaceFile(
            "shared/capabilityrecipe/capability_runtime.v1.json",
        ).readText().trim()
        val runtime = registry.removeSuffix("}") +
            ",\"revision\":\"production-snapshot\",\"generatedAt\":\"2026-08-13T00:00:00Z\"}"

        val providers = buildJsonObject {
            models.values.groupBy { it["providerKind"]!!.jsonPrimitive.content }
                .forEach { (kind, entries) ->
                    put(
                        kind,
                        buildJsonObject {
                            put(
                                "resolveMap",
                                buildJsonObject {
                                    entries.forEach { entry ->
                                        val id = entry["modelId"]!!.jsonPrimitive.content
                                        put(id, id)
                                    }
                                },
                            )
                            put(
                                "models",
                                buildJsonObject {
                                    entries.forEach { entry ->
                                        val id = entry["modelId"]!!.jsonPrimitive.content
                                        put(
                                            id,
                                            buildJsonObject {
                                                put("canonicalModelId", id)
                                                put("transport", entry["transport"]!!.jsonPrimitive.content)
                                                put("capabilities", entry["capabilities"]!!.jsonArray)
                                                put("profiles", entry["profiles"]!!.jsonObject)
                                                put("capabilityControls", entry["capabilityControls"]!!.jsonObject)
                                            },
                                        )
                                    }
                                },
                            )
                        },
                    )
                }
        }
        val reasoningProfiles = snapshot["profiles"]!!.jsonObject["reasoning"]!!.jsonObject
        val payload = buildJsonObject {
            put("version", JsonPrimitive(1))
            put("profiles", buildJsonObject { put("reasoning", reasoningProfiles) })
            put("providers", providers)
        }.toString()
        // capabilityRuntime is injected as raw registry text, so buildJsonObject never gets to
        // reorder it.
        val withRuntime = payload.removeSuffix("}") + ",\"capabilityRuntime\":$runtime}"

        val client = MetadataClient()
        client.loadNetworkPayloadForTesting(withRuntime, "etag-production-snapshot")
        return client
    }

    private fun verdict(
        client: MetadataClient,
        key: String,
        capability: String,
    ): CapabilityControlResolution.Verdict {
        val entry = models.getValue(key)
        val kind = ProviderKind.entries.first { it.rawValue == entry["providerKind"]!!.jsonPrimitive.content }
        val modelId = entry["modelId"]!!.jsonPrimitive.content
        return CapabilityControlResolution.resolve(
            provider = Provider(id = "p", kind = kind),
            model = AIModel(id = modelId, name = modelId),
            capability = capability,
            metadata = client,
        )
    }

    /** Every model in the slice must have an expected verdict; cherry-picking a handful of
     * capable models is precisely what let the previous round through. */
    @Test
    fun `covers every model in the shared slice - no cherry-picking`() {
        assertEquals(models.keys.sorted(), expectedVerdicts.keys.sorted())
    }

    /**
     * The full per-model matrix. Of the hundreds of models in the published catalog, roughly
     * half cannot emit a web-search tool at all, so most users land on the empty side. Models
     * with no capability are first-class cases here, not an afterthought.
     */
    @Test
    fun `resolves web and reasoning for every production model exactly as the shared table says`() {
        val client = metadataClient()
        listOf("web", "reasoning").forEach { capability ->
            var available = 0
            var unavailable = 0
            models.keys.forEach { key ->
                val expected = expectedVerdicts.getValue(key)[capability]!!.jsonObject
                val actual = verdict(client, key, capability)
                assertEquals(
                    "$key/$capability state",
                    expected["state"]!!.jsonPrimitive.content,
                    actual.state,
                )
                assertEquals(
                    "$key/$capability viaLegacyProfile",
                    expected["viaLegacyProfile"]!!.jsonPrimitive.content.toBoolean(),
                    actual.viaLegacyProfile,
                )
                expected["intents"]?.let { intents ->
                    assertEquals(
                        "$key/$capability intents",
                        intents.jsonArray.map { it.jsonPrimitive.content },
                        actual.intents,
                    )
                }
                // The shared criterion behind the badge and the control has to be decided by
                // state alone; nowhere else may add a condition of its own.
                if (actual.isAvailable) available += 1 else unavailable += 1
            }
            // Empty states are first-class: if the slice is ever trimmed down to only the
            // capable models, these two go red.
            assertTrue("$capability: the available side must not be empty", available > 0)
            assertTrue("$capability: the unavailable side must not be empty", unavailable > 0)
        }
    }

    /**
     * "Automatic configuration exists but there is no ladder to choose from" is a first-class
     * state, not a broken auto_available.
     *
     * Observed in the published catalog: `openAI/gpt-5-pro` accepts only high and
     * `gpt-5.2-chat-latest` only medium, and the catalog deliberately publishes no
     * `availableIntents` for either. The client must neither downgrade them to unavailable nor
     * invent tiers of its own; inventing tiers is what produces the dead end where the panel
     * says the control is available and then offers nothing to pick.
     */
    @Test
    fun `an automatic reasoning control with an empty ladder stays available and stays empty`() {
        val client = metadataClient()
        listOf("openAI/gpt-5-pro", "openAI/gpt-5.2-chat-latest").forEach { key ->
            val entry = models.getValue(key)
            val reasoning = entry["capabilityControls"]!!.jsonObject["reasoning"]!!.jsonObject
            // Self-evidencing premise: the slice really is shaped like this, the assertions
            // below are not assuming it.
            assertEquals("$key slice state", "auto_available", reasoning["state"]!!.jsonPrimitive.content)
            assertNull("$key slice carries no availableIntents", reasoning["availableIntents"])
            assertTrue(
                "$key slice carries no legacy reasoning profile",
                entry["profiles"]!!.jsonObject["reasoning"] is JsonNull,
            )

            val actual = verdict(client, key, "reasoning")
            assertEquals("auto_available", actual.state)
            assertEquals(emptyList<String>(), actual.intents)
            assertTrue(actual.isAvailable)
        }
    }

    /**
     * Only an exact official recipe is trusted. Even when a historical snapshot still carries a
     * legacy web profile, unknown has to stay unknown; it may not be guessed into available
     * from the provider, the model id or the profile.
     */
    @Test
    fun `an unknown v2 control stays unknown and never falls back to a legacy profile`() {
        val client = metadataClient()
        val actual = verdict(client, "qwen/qwen3-max", "web")
        assertEquals("unknown", actual.state)
        assertTrue(!actual.viaLegacyProfile)
        assertEquals("official_source_insufficient", actual.reasonCode)
    }

    /** The reverse assertion: when v2 says outright that a capability is unsupported, the
     * legacy fallback may not lift it back to available. */
    @Test
    fun `an explicitly unavailable control is never lifted by a legacy profile`() {
        val client = metadataClient()
        val actual = verdict(client, "qwen/qwen3.6-plus", "web")
        assertEquals("unavailable", actual.state)
        assertTrue(!actual.viaLegacyProfile)
    }

    private fun workspaceFile(relative: String): File {
        var dir: File? = File("").absoluteFile
        while (dir != null) {
            val candidate = File(dir, relative)
            if (candidate.exists()) return candidate
            dir = dir.parentFile
        }
        throw IllegalStateException("cannot locate $relative")
    }
}

File: web/apps/app/lib/core/chat/production-capability-snapshot.test.ts
import { readFileSync } from 'node:fs';
import { resolve } from 'node:path';
import { describe, expect, it, vi } from 'vitest';

/**
 * Drives the real verdict functions with the **metadata slice production actually serves**.
 *
 * Why not another set of hand-written fixtures: the previous round of verification was all "unit
 * tests pass / curl shows the right field / deploy returned 200", none of which says anything about
 * what the user sees on screen, and every test built its own data to make its own assertion hold.
 * The state/profile combinations here are lifted verbatim from the live metadata response, so the
 * assertions are about models users really hit that day.
 *
 * The slice itself lives in `shared/model-contracts/`, so one snapshot of production truth feeds the
 * real verdict function of each client and whichever one drifts turns red first.
 * The expected values live in that same slice (`expectedVerdicts`) - keeping a separate copy per
 * client would lock nothing at all.
 */

// vitest starts with cwd = apps/app; three levels up (apps -> web -> repository root) reaches shared/.
const snapshot = JSON.parse(
  readFileSync(
    resolve(process.cwd(), '../../..', 'shared/model-contracts/production-capability-snapshot.json'),
    'utf-8',
  ),
) as {
  profiles: { reasoning: Record<string, { levels: string[] }> };
  models: Record<string, unknown>;
  expectedVerdicts: Record<string, unknown>;
};

const runtime = vi.hoisted(() => ({
  catalog: null as any,
  reasoningLevels: [] as string[],
}));

/**
 * Every recipeRef that appears in the slice must be registered in the runtime.
 *
 * This mock used to pass `recipes: {}` with a comment claiming it only had to be non-empty, which
 * made the suite falsely green: `resolveControl` looks recipeRef up in `recipes`, and a miss is
 * judged `dangling_recipe_ref` and degrades the whole entry to unknown, so every auto_available
 * control carrying a recipeRef fell through to the legacy path - the assertion about "not consulting
 * the legacy profile" was in fact passing via the legacy profile. Production really does publish
 * these recipes (that is why the controls carry a recipeRef), so filling them in is the true shape.
 */
const publishedRecipes = Object.fromEntries(
  Object.values(snapshot.models as Record<string, { capabilityControls: Record<string, { recipeRef?: string }> }>)
    .flatMap((entry) => Object.values(entry.capabilityControls))
    .map((control) => control.recipeRef)
    .filter((ref): ref is string => typeof ref === 'string')
    .map((ref) => [ref, {}]),
);

vi.mock('../metadata/metadata-client', () => ({
  // The production capabilityRuntime is published; recipe compilation is locked by the
  // request-compiler contract tests, so all that is needed here is a resolvable recipeRef, letting
  // auto_available controls take the v2 branch for real.
  getCapabilityRuntime: () => ({ recipes: publishedRecipes, sourceIndex: {}, controlDefinitions: {} }),
  getModelTransport: () => undefined,
  resolveCatalogModel: () => runtime.catalog,
  getDeclaredReasoningLevels: (name?: string | null) => (
    name ? snapshot.profiles.reasoning[name]?.levels ?? [] : []
  ),
}));

// eslint-disable-next-line import/first -- vi.mock must be evaluated before the module under test
import { presentCapabilityControl } from './capability-control-presentation';

type SnapshotModel = {
  providerKind: string;
  modelId: string;
  capabilities: string[];
  profiles: { reasoning: string | null; webSearch: string | null };
  capabilityControls: Record<string, { state?: string; reasonCode?: string }>;
  transport: string;
};

type ExpectedVerdict = {
  state: string;
  viaLegacyProfile: boolean;
  /** Omitted means the exact levels are unconstrained; only state and viaLegacyProfile are checked. */
  intents?: string[];
};

const models = snapshot.models as unknown as Record<string, SnapshotModel>;
const expectedVerdicts = Object.fromEntries(
  Object.entries(snapshot.expectedVerdicts).filter(([key]) => key !== '$comment'),
) as Record<string, Record<'web' | 'reasoning', ExpectedVerdict>>;

const CAPABILITIES = ['web', 'reasoning'] as const;

function verdict(key: string, capability: 'web' | 'reasoning') {
  const entry = models[key];
  const provider = { id: 'p', kind: entry.providerKind, status: { kind: 'connected' } } as any;
  runtime.catalog = {
    capabilities: entry.capabilities,
    profiles: { webSearch: entry.profiles.webSearch, reasoning: entry.profiles.reasoning },
    transport: entry.transport,
  };
  const model = {
    id: entry.modelId,
    capabilities: entry.capabilities,
    capabilityControls: entry.capabilityControls,
  } as any;
  return presentCapabilityControl(provider, model, capability);
}

describe('production capability snapshot', () => {
  it('keeps Qwen unknown when the exact v2 recipe is not published yet', () => {
    // Live: capabilityControls.web = unknown/official_source_insufficient. Even though the existing
    // profiles.webSearch still carries qwen_web, a new client must not treat that as an automatic
    // configuration. The valid final body for ordinary chat is locked separately by the production
    // builder contract tests.
    expect(verdict('qwen/qwen3-max', 'web')).toMatchObject({
      state: 'unknown',
      viaLegacyProfile: false,
    });
  });

  it('keeps Web off where the recipe positively says this transport cannot do it', () => {
    // Live: unavailable/transport_not_supported - the official note says Responses API only.
    expect(verdict('qwen/qwen3.6-plus', 'web').state).toBe('unavailable');
  });

  it('honours a published v2 recipe without consulting any legacy profile', () => {
    for (const [key, capability] of [['zhipu/glm-4.6', 'web'], ['anthropic/claude-fable-5', 'reasoning']] as const) {
      const presented = verdict(key, capability);
      expect(presented.state, `${key}/${capability}`).toBe('auto_available');
      expect(presented.viaLegacyProfile, `${key}/${capability}`).toBe(false);
    }
  });

  it('does not invent Web for a model that has neither a control nor a profile', () => {
    expect(verdict('anthropic/claude-fable-5', 'web').state).toBe('unavailable');
  });

  it('leaves an explicitly unsupported reasoning control unavailable', () => {
    // Live, 128 openRouter models are unavailable/upstream_parameter_not_declared with no legacy
    // reasoning profile - the fallback must not promote them to available.
    expect(verdict('openRouter/anthropic/claude-opus-4.5', 'reasoning').state).toBe('unavailable');
  });

  it('exposes legacy reasoning tiers in v2 intent vocabulary', () => {
    const presented = verdict('zhipu/glm-4.6', 'reasoning');
    expect(presented.state).toBe('auto_available');
  });

  // -- Full per-model matrix ------------------------------------------------
  //
  // Testing only "a few hand-picked models that have capabilities" is exactly what went wrong last
  // time: 362 of the 764 live models cannot emit a single web search tool, and most users land on
  // the empty side. This runs the real verdict for **every** model in the slice, against the
  // expectedVerdicts the slice carries.

  it('covers every model in the shared slice — no cherry-picking', () => {
    expect(Object.keys(expectedVerdicts).sort()).toEqual(Object.keys(models).sort());
  });

  it.each(CAPABILITIES)('resolves %s for every production model exactly as the shared table says', (capability) => {
    let available = 0;
    let unavailable = 0;
    for (const key of Object.keys(models)) {
      const expected = expectedVerdicts[key]![capability];
      const presented = verdict(key, capability);
      expect(presented.state, `${key}/${capability} state`).toBe(expected.state);
      expect(presented.viaLegacyProfile, `${key}/${capability} viaLegacyProfile`).toBe(expected.viaLegacyProfile);
      if (expected.intents) {
        expect([...presented.availableIntents], `${key}/${capability} intents`).toEqual(expected.intents);
      }
      // The shared test for badge and control must be decided by state alone; nowhere may add a rule of its own.
      const usable = presented.state === 'auto_available' || presented.state === 'managed_only';
      usable ? (available += 1) : (unavailable += 1);
    }
    // The empty side is a first-class case: if the slice is ever trimmed down to only the capable
    // models, these two turn red.
    expect(available, `${capability}: available side must not be empty`).toBeGreaterThan(0);
    expect(unavailable, `${capability}: unavailable side must not be empty`).toBeGreaterThan(0);
  });

  // -- "Automatic configuration but no ladder" is a first-class state, not a broken auto_available --
  //
  // Production evidence: reasoning for openAI/gpt-5-pro and gpt-5.2-chat-latest is auto_available
  // with an empty availableIntents and no legacy reasoning profile. That is deliberate on the server
  // side (the pro variant rejects low, and these two single-value models accept only high and medium
  // respectively; with no selectable level there is no point drawing a button that must fail). The
  // client must not downgrade this to unavailable, nor invent levels - that leads straight to "the
  // panel says available and offers no level at all".
  it.each(['openAI/gpt-5-pro', 'openAI/gpt-5.2-chat-latest'])(
    'keeps %s reasoning auto_available with an honestly empty intent ladder',
    (key) => {
      const entry = models[key]!;
      // Prove the premise: the slice itself has this shape, rather than the assertion assuming it.
      expect(entry.capabilityControls.reasoning!.state, `${key} slice state`).toBe('auto_available');
      expect((entry.capabilityControls.reasoning as { availableIntents?: string[] }).availableIntents,
        `${key} slice has no availableIntents`).toBeUndefined();
      expect(entry.profiles.reasoning, `${key} slice has no legacy reasoning profile`).toBeNull();

      const presented = verdict(key, 'reasoning');
      expect(presented.state).toBe('auto_available');
      expect(presented.availableIntents).toEqual([]);
      expect(presented.viaLegacyProfile).toBe(false);
    },
  );
});
